### ISSUE TITLE
regression in training accuracy with VGG16 application

### ISSUE THREAD

**Issue by jjallaire** [keras-team/keras#8792]
Please make sure that the boxes below are checked before you submit your issue. If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

When running under both Keras 2.1.2 as well as Keras from GitHub master the validation accuracy achieved in the dogs vs. cats example in Chapter 5.3 of the DL w/Python book (https://github.com/fchollet/deep-learning-with-python-notebooks/blob/master/5.3-using-a-pretrained-convnet.ipynb) is ~ 90%.

However, the validation accuracy when running under Keras 2.0.9 is ~ 96%.

The code (as well as the cats vs. dogs images) required to reproduce this is in the following GitHub repo (the script is also inline below):

https://github.com/jjallaire/keras-vgg16-accuracy

One other noteworthy thing is that with Keras 2.1.2 case training takes ~ 22 seconds/epoch whereas with Keras 2.0.9 it's ~ 50 seconds per epoch (this on an Amazon p2.xlarge instance).

```python
import os
import keras
from keras import models
from keras import layers
from keras import optimizers
from keras.applications import VGG16
from keras.preprocessing.image import ImageDataGenerator


base_dir = 'cats_and_dogs_small'
train_dir = os.path.join(base_dir, 'train')
validation_dir = os.path.join(base_dir, 'validation')
test_dir = os.path.join(base_dir, 'test')


conv_base = VGG16(weights='imagenet',
                  include_top=False,
                  input_shape=(150, 150, 3))
                  
model = models.Sequential()
model.add(conv_base)
model.add(layers.Flatten())
model.add(layers.Dense(256, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

conv_base.trainable = False


train_datagen = ImageDataGenerator(
      rescale=1./255,
      rotation_range=40,
      width_shift_range=0.2,
      height_shift_range=0.2,
      shear_range=0.2,
      zoom_range=0.2,
      horizontal_flip=True,
      fill_mode='nearest')

test_datagen = ImageDataGenerator(rescale=1./255)

train_generator = train_datagen.flow_from_directory(
        train_dir,
        target_size=(150, 150),
        batch_size=20,
        class_mode='binary')

validation_generator = test_datagen.flow_from_directory(
        validation_dir,
        target_size=(150, 150),
        batch_size=20,
        class_mode='binary')


model.compile(loss='binary_crossentropy',
              optimizer=optimizers.RMSprop(lr=2e-5),
              metrics=['acc'])

history = model.fit_generator(
      train_generator,
      steps_per_epoch=100,
      epochs=30,
      validation_data=validation_generator,
      validation_steps=50,
      verbose=1)
```

**Comment by jjallaire** [comment]
I am also seeing the ~ 90% validation accuracy in Keras 2.1.0 so it appears that a change between 2.0.9 and 2.1.0 precipitated this.

**Comment by fchollet** [comment]
Please try to run a single `train_on_batch` from a known state (e.g. freshly initialized optimizer and imagenet weights, same input array and target array, no preprocessing/augmentation) and compare the outputs with 2.0.9 and 2.1.2.

**Comment by jjallaire** [comment]
From git bisect I found that this commit is the one where the behavior starts deviating: https://github.com/keras-team/keras/commit/c25fa38deb4efc5445f64af3ec17eae0eb660d2f

**Comment by fchollet** [comment]
Do you get the old behavior by removing `conv_base.trainable = False`?

**Comment by jjallaire** [comment]
Yes, if I remove `conv_base.trainable = False` when running under 2.1.0 then I get the old behavior! (higher accuracy, ~ 50 seconds / epoch training time)

**Comment by fchollet** [comment]
The current behavior is correct. Past behavior was a hard-to-detect bug. There's no easy fix here.

**Comment by jjallaire** [comment]
Okay, very well, glad we figured this out!
